src/uffd_handler/src/handler.rs

```diff
@@ -1,59 +1,159 @@
+use std::os::unix::net::UnixStream;
 
-mod common;
+use crate::{
+    create_mem_regions, parse_unix_stream, GuestRegionUffdMapping, MemPageState, MemRegion,
+};
+use userfaultfd::Uffd;
+use utils::get_page_size;
 
-use std::os::unix::io::AsRawFd;
+pub struct UffdPfHandler {
+    mem_regions: Vec<MemRegion>,
+    backing_buffer: *const u8,
+    pub uffd: Uffd,
+    // Not currently used but included to demonstrate how a page fault handler can
+    // fetch Firecracker's PID in order to make it aware of any crashes/exits.
+    _firecracker_pid: u32,
+}
 
-use common::{create_pf_handler, MemPageState};
+impl UffdPfHandler {
+    pub fn from_unix_stream(stream: UnixStream, data: *const u8, size: usize) -> Self {
+        let (file, msg_body) = parse_unix_stream(&stream);
+        // SAFETY: Safe because it wraps the Uffd object around the valid raw file descriptor.
+        let uffd = unsafe { Uffd::from_raw_fd(file.into_raw_fd()) };
 
-fn main() {
-    let mut uffd_handler = create_pf_handler();
+        // Create guest memory regions from mappings received from Firecracker process.
+        let mappings = serde_json::from_str::<Vec<GuestRegionUffdMapping>>(&msg_body)
+            .expect("Cannot deserialize memory mappings.");
+        let memsize: usize = mappings.iter().map(|r| r.size).sum();
+        let mem_regions = create_mem_regions(&mappings);
 
-    let mut pollfd = libc::pollfd {
-        fd: uffd_handler.uffd.as_raw_fd(),
-        events: libc::POLLIN,
-        revents: 0,
-    };
-    // Loop, handling incoming events on the userfaultfd file descriptor.
-    loop {
-        // See what poll() tells us about the userfaultfd.
-        let nready = unsafe { libc::poll(&mut pollfd, 1, -1) };
+        // Make sure memory size matches backing data size.
+        assert_eq!(memsize, size);
 
-        if nready == -1 {
-            panic!("Could not poll for events!")
+        // Get credentials of Firecracker process sent through the stream.
+        let creds: libc::ucred = get_peer_process_credentials(stream);
+
+        Self {
+            mem_regions,
+            backing_buffer: data,
+            uffd,
+            _firecracker_pid: creds.pid as u32,
         }
+    }
 
-        let revents = pollfd.revents;
+    pub fn update_mem_state_mappings(&mut self, start: u64, end: u64, state: &MemPageState) {
+        for region in self.mem_regions.iter_mut() {
+            for (key, value) in region.page_states.iter_mut() {
+                if key >= &start && key < &end {
+                    *value = state.clone();
+                }
+            }
+        }
+    }
 
-        println!(
-            "poll() returns: nready = {}; POLLIN = {}; POLLERR = {}",
-            nready,
-            revents & libc::POLLIN,
-            revents & libc::POLLERR,
-        );
+    fn populate_from_file(&self, region: &MemRegion) -> (u64, u64) {
+        let src = self.backing_buffer as u64 + region.mapping.offset;
```

Comment:
It seems like `self.backing_buffer` starts life as a `*mut c_void` returned from `mmap`, it casted to `*const u8` when stored in `PageFaultHandler::backing_buffer`, is converted to a `u64` here, just to be casted back to `*const c_void` below. Could we have it be a `*const c_void` throughout its entire lifecycle, and do the pointer arithmetic here via [`ptr::add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.add)?

Reply:
Mh, the reason I'd be in favor of `ptr::add` is that even if we avoid it (and thus `unsafe`), we'd still have to uphold all the invariants it requires, but without `unsafe` we get lulled into a false sense of security in the sense that "oh, nothing can go wrong" - although I'm also not sure how much of those invariants even imply here, since we don't get the pointer from rust's allocator, so am happy to leave things the way they are :)